[mlir] Decouple enum generation from attributes, adding EnumInfo and EnumCase

krzysz00 opened the pull request:
This commit pulls apart the inherent attribute dependence of classes like EnumAttrInfo and EnumAttrCase, factoring them out into simpler EnumCase and EnumInfo variants. This allows specifying the cases of an enum without needing to make the cases, or the EnumInfo itself, a subclass of SignlessIntegerAttrBase.

The existing classes are retained as subclasses of the new ones, both for backwards compatibility and to allow attribute-specific information.

In addition, the new BitEnum class changes its default printer/parser behavior: cases when multiple keywords appear, like having both nuw and nsw in overflow flags, will no longer be quoted by the operator<<, and the FieldParser instance will now expect multiple keywords. All instances of BitEnumAttr retain the old behavior.

matthias-springer reviewed (commented):
Is it possible to split this up into multiple smaller PRs?

krzysz00 commented:
@matthias-springer I can split the EnumProp changes into their own PR, not I don't think there split points other than that

krzysz00 committed:
[mlir] Decouple enum generation from attributes, adding EnumInfo and EnumCase

This commit pulls apart the inherent attribute dependence of classes
like EnumAttrInfo and EnumAttrCase, factoring them out into simpler
EnumCase and EnumInfo variants. This allows specifying the cases of an
enum without needing to make the cases, or the EnumInfo itself, a
subclass of SignlessIntegerAttrBase.

The existing classes are retained as subclasses of the new ones, both
for backwards compatibility and to allow attribute-specific
information.

In addition, the new BitEnum class changes its default printer/parser
behavior: cases when multiple keywords appear, like having both nuw
and nsw in overflow flags, will no longer be quoted by the operator<<,
and the FieldParser instance will now expect multiple keywords. All
instances of BitEnumAttr retain the old behavior.

krzysz00 commented:
Have split the EnumProp thing into #132349, hopefully this improves reviewability

matthias-springer commented:
I'm not super familiar with this part of the code base. `Operations.md` talks about `EnumAttrCase`, should it also be updated?

matthias-springer commented on mlir/tools/mlir-tblgen/EnumsGen.cpp:
```diff
@@ -128,8 +143,95 @@ namespace llvm {
+  struct FieldParser;
+
+  template<>
+  struct FieldParser<{0}, {0}> {{
```
How was this handled before? I expected that this is would replace another `FieldParser` implementation.

krzysz00 replied:
There's a FieldParser implementation above that's the old-style bit enum. It's handled by always quoting bit enum values that have more than one bit set.

Usually, these FieldParser values were very much unused, as there are separate parser implementatinos both in `EnumParameter` and in other bits of `mlir-tblgen` that mean this code was basically never called.

matthias-springer commented:
Thx for breaking up the PR. It's still quite large, maybe this could also be a separate one:
```
In addition, the new BitEnum class changes its default printer/parser behavior: cases when multiple keywords appear, like having both nuw and nsw in overflow flags, will no longer be quoted by the operator<<, and the FieldParser instance will now expect multiple keywords. All instances of BitEnumAttr retain the old behavior.
```

Or maybe the part that moves code to `EnumInfo.h`.

In my experience, anything that's larger than 100 loc is taking a quite long time to get reviewed...

krzysz00 commented on mlir/tools/mlir-tblgen/EnumsGen.cpp:
```diff
@@ -66,20 +67,34 @@ static void emitEnumClass(const Record &enumDef, StringRef enumName,
+  llvm::raw_string_ostream caseListOs(casesList);
+  caseListOs << "[";
+  for (auto [index, caseVal] : llvm::enumerate(cases)) {
+    if (index != 0)
```
Looks like I can, yeah

krzysz00 commented:
Have updated documentation.

Do folks have a strong opinion on if I should move the new parser stuff into a separate commit?

makslevental commented:
> enum binding code in IREE?

nah I doubt it - the handwritten bindings in [compiler/bindings/python/IREECompilerDialectsModule.cpp](https://github.com/makslevental/iree/blob/main/compiler/bindings/python/IREECompilerDialectsModule.cpp#L25-L46) is ... handwritten (so unaffected). Besides that I don't think (don't remember) IREE (or tuner...) is using these autogenerated bindings (`*_enums_gen.py`) and even so I believe (since I trust @krzysz00, though I haven't looked closely) this change won't affect those either - at minimum the in-tree tests here (for the `*_enums_gen.py` bindings) passed so we're probably good.

krzysz00 commented:
Re IREE bindings: If nothing else, this change should be NFC for anything that uses `EnumAttrInfo`/`EnumAttr`

Whether the Python bindings for non-attribute enums are good / how Python binds non-attribute properties ... I'll cross that bridge when I get there

makslevental commented:
> how Python binds non-attribute properties

It doesn't - the lines you touched (there-abouts) are the only place where Enum bindings are automatically emitted (and "handwritten" bindings should not be affected in any way).
